src/node/stage/mod.rs

```diff
@@ -8,39 +8,373 @@
 
 mod approved;
 mod bootstrapping;
+mod comm;
 mod joining;
 
-pub(crate) use self::{
-    approved::{Approved, RelocateParams},
-    bootstrapping::{Bootstrapping, JoinParams},
-    joining::Joining,
+use self::{approved::Approved, bootstrapping::Bootstrapping, comm::Comm, joining::Joining};
+use crate::{
+    consensus::{self, Proof, Proven},
+    error::{Error, Result},
+    event::{Connected, Event},
+    id::{FullId, P2pNode},
+    location::{DstLocation, SrcLocation},
+    messages::{Message, Variant},
+    network_params::NetworkParams,
+    rng::MainRng,
+    section::{EldersInfo, MemberInfo, SectionKeyShare, SectionProofChain, SharedState, MIN_AGE},
+    timer::Timer,
+    TransportConfig,
 };
+use bytes::Bytes;
+use qp2p::IncomingConnections;
+use serde::Serialize;
+use std::{iter, net::SocketAddr, sync::Arc};
+use tokio::sync::mpsc;
+use xor_name::Prefix;
 
 #[cfg(feature = "mock")]
 pub use self::{bootstrapping::BOOTSTRAP_TIMEOUT, joining::JOIN_TIMEOUT};
 
-// Type to represent the various stages a node goes through during its lifetime.
+// Maximum number of events to be buffered internally, when the buffer is full
+// no new events can be generated by this crate
+// TODO: if external connections or messages are arriving when
+// the buffer is full, they need to be rejected.
+const MAX_EVENTS_BUFFERED: usize = 1024;
```

Comment:
We wouldn't need this if we used unbounded channels: https://docs.rs/tokio/0.2.22/tokio/sync/mpsc/fn.unbounded_channel.html